optuna/samplers/_tpe/sampler.py

```diff
@@ -720,18 +720,26 @@ def _split_observation_pairs(
 
 
 def _calculate_weights_below_for_multi_objective(
-    loss_vals: List[Tuple[float, List[float]]],
-    indices: np.ndarray,
-    violations: Optional[List[float]],
+    study: Study,
+    below_trials: list[FrozenTrial],
+    constraints_func: Callable[[FrozenTrial], Sequence[float]] | None,
 ) -> np.ndarray:
-    if violations is None:
-        feasible_mask = np.ones(len(indices), dtype=bool)
-    else:
+    loss_vals = []
+    feasible_mask = np.ones(len(below_trials), dtype=bool)
+    for i, trial in enumerate(below_trials):
         # Hypervolume contributions are calculated only using feasible trials.
-        feasible_mask = np.array(violations, dtype=float)[indices] == 0
-
-    # Multi-objective TPE does not support pruning, so it ignores the ``step``.
-    lvals = np.asarray([v for _, v in loss_vals])[indices[feasible_mask]]
+        if constraints_func is not None:
+            if any(constraint > 0 for constraint in constraints_func(trial)):
+                feasible_mask[i] = False
+                continue
+        values = []
+        for value, direction in zip(trial.values, study.directions):
+            if direction == StudyDirection.MINIMIZE:
+                values.append(value)
+            else:
+                values.append(-value)
```

Comment:
It would be good to add a maximization study in tests for coverage, but it is not necessary condition to merge.